### ISSUE TITLE
"tox.tox_env.errors.Recreate: requirements removed" if requirements change

### ISSUE THREAD

**Issue by fohrloop** [sclabs/tox-extras#2]
Thanks for the great tool! It's such a time saver.

I noticed that if you have to change any dependencies in the extras in your `pyproject.toml`, the tox run fails with an Exception similar to this

```
Traceback (most recent call last):
  File "/path/to/myproj/.tox/.tox/lib/python3.10/site-packages/tox/session/cmd/run/single.py", line 48, in _evaluate
    code, outcomes = run_commands(tox_env, no_test)
  File "/path/to/myproj/.tox/.tox/lib/python3.10/site-packages/tox/session/cmd/run/single.py", line 79, in run_commands
    MANAGER.tox_before_run_commands(tox_env)
  File "/path/to/myproj/.tox/.tox/lib/python3.10/site-packages/tox/plugin/manager.py", line 88, in tox_before_run_commands
    self.manager.hook.tox_before_run_commands(tox_env=tox_env)
  File "/path/to/myproj/.tox/.tox/lib/python3.10/site-packages/pluggy/_hooks.py", line 501, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/path/to/myproj/.tox/.tox/lib/python3.10/site-packages/pluggy/_manager.py", line 119, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/path/to/myproj/.tox/.tox/lib/python3.10/site-packages/pluggy/_callers.py", line 138, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/path/to/myproj/.tox/.tox/lib/python3.10/site-packages/pluggy/_callers.py", line 102, in _multicall
    res = hook_impl.function(*args)
  File "/path/to/myproj/.tox/.tox/lib/python3.10/site-packages/tox_extras.py", line 57, in tox_before_run_commands
    tox_env.installer.install(extras[extra], f"tox_extras_{extra}", "deps")
  File "/path/to/myproj/.tox/.tox/lib/python3.10/site-packages/tox/tox_env/python/pip/pip_install.py", line 92, in install
    self._install_requirement_file(arguments, section, of_type)
  File "/path/to/myproj/.tox/.tox/lib/python3.10/site-packages/tox/tox_env/python/pip/pip_install.py", line 139, in _install_requirement_file
    raise Recreate(msg)
tox.tox_env.errors.Recreate: requirements removed: isort~=5.11.0
  check: FAIL code 2 (0.06 seconds)
  evaluation failed :( (0.18 seconds)
```

The workaround is to remove the `.tox` folder and run again.

**Comment by thomasgilgenast** [comment]
Thanks for reporting this bug. I can reproduce this by running `tox` from `example/`, then applying this diff

```diff
diff --git a/example/pyproject.toml b/example/pyproject.toml
index 89d21e7..3e0bbbb 100644
--- a/example/pyproject.toml
+++ b/example/pyproject.toml
@@ -7,4 +7,4 @@ name = "example"
 version = "0.0.0"

 [project.optional-dependencies]
-lint = ["black==23.3.0"]
+lint = ["black==24.2.0"]
```

then running `tox` again.

I think this is related to the fact that tox-extras somehow bypasses or messes with the expectations of how tox caches testenvs. The dependencies that get installed in the testenv via the `tox_extras` key are not taken into account when tox decides whether or not the testenv needs to be recreated at the beginning of a tox run.

Ideally, I should probably look for a way to make tox-extras somehow include the packages implied by `tox_extras` into the tox testenv cache key. But in the short run, I came up with the following PR: #3 If you'd like to test drive it, you can do

    pip install git+https://github.com/sclabs/tox-extras@allow-changing-deps-in-extras

and temporarily unpin or remove `tox-extras` from the `tox.ini` `[tox]` `requires` section.

In a very limited test on my end, this led to the following console output:

```
lintonly: recreate env because requirements removed: black==23.3.0
lintonly: remove tox env folder C:\Users\thoma\git\tox-extras\example\.tox\lintonly
lintonly: install_deps> python -I -m pip install black==24.2.0
lintonly: commands[0]> black --check example.py
...
```

and the tox run succeeds.

I am seeing some errors in CI on this PR branch related to

    ModuleNotFoundError: No module named 'pkg_resources'

and I am working on simplifying the test setup to try to avoid the hacky approach I take there, but the core functionality seems to work already in my limited testing.

**Comment by fohrloop** [comment]
Thanks for the swift response! I tested with 

```
pip install git+https://github.com/sclabs/tox-extras@allow-changing-deps-in-extras
```

and changed the requirements in the extras in pyproject.toml and did not get any errors. I saw from the logs that this time the env was recreated automatically.

**Comment by thomasgilgenast** [comment]
Thank you for test driving the PR! I've merged it and the fix is now published in tox-extras v0.0.2

https://pypi.org/project/tox-extras/0.0.2/#files
